### PR TITLE
[CLOUDSTACK-9643] Now returning os info with the list snapshot response

### PR DESCRIPTION
**Pull request by ProjectMoon**
This commit adds the ID and display name of the OS on the volume.

### CODE REVIEW

**Comment by ProjectMoon** [comment]
Rebased to latest master.

**Comment by rhtyd** [comment]
@blueorangutan package
@ProjectMoon can you push -f to rekick Travis?

File: server/src/com/cloud/api/ApiResponseHelper.java [hunk 1]
```diff
@@ -157,6 +156,7 @@
 import org.apache.cloudstack.usage.Usage;
 import org.apache.cloudstack.usage.UsageService;
 import org.apache.cloudstack.usage.UsageTypes;
+import org.apache.commons.collections.CollectionUtils;
```
Comment (jburwell) [inline server/src/com/cloud/api/ApiResponseHelper.java]: Why was this import added?  I don't see it used anywhere in the patch.

File: server/src/com/cloud/api/ApiResponseHelper.java [hunk 2]
```diff
@@ -479,6 +479,17 @@ public SnapshotResponse createSnapshotResponse(Snapshot snapshot) {
             if (zone != null) {
                 snapshotResponse.setZoneId(zone.getUuid());
             }
+
+            if (volume.getVolumeType().equals(Volume.Type.ROOT)) {
```
Comment (jburwell) [inline server/src/com/cloud/api/ApiResponseHelper.java]: Enumerated values are guaranteed to be singletons.   Therefore, the ``==`` operator deterministically performs both reference and value equality checks.  It is also preferred because it avoid an NPE in the event that ``volume.getVolumeType()`` returns null.

File: server/src/com/cloud/api/ApiResponseHelper.java [hunk 3]
```diff
@@ -479,6 +479,17 @@ public SnapshotResponse createSnapshotResponse(Snapshot snapshot) {
...
+                VMInstanceVO instance = ApiDBUtils.findVMInstanceById(volume.getInstanceId());
+                if (instance != null) {
+                    GuestOS guestOs = ApiDBUtils.findGuestOSById(instance.getGuestOSId());
```
Comment (jburwell) [inline server/src/com/cloud/api/ApiResponseHelper.java]: It feels like lines 484 & 486 could be combined into a single join exposed by a method on ``VolumeDao``.  Do you agree?

File: api/src/org/apache/cloudstack/api/response/SnapshotResponse.java [hunk 4]
```diff
@@ -199,4 +207,20 @@ public boolean isRevertable() {
     public void setRevertable(boolean revertable) {
         this.revertable = revertable;
     }
-}
+
+    public String getOsTypeId() {
```
Comment (koushik-das) [inline api/src/org/apache/cloudstack/api/response/SnapshotResponse.java]: The get* methods are not used, please remove.

File: api/src/org/apache/cloudstack/api/response/SnapshotResponse.java [hunk 5]
```diff
@@ -102,6 +102,14 @@
     @Param(description = "indicates whether the underlying storage supports reverting the volume to this snapshot")
     private boolean revertable;
 
+    @SerializedName(ApiConstants.OS_TYPE_ID)
+    @Param(description = "id of the os on volume")
```
Comment (koushik-das) [inline api/src/org/apache/cloudstack/api/response/SnapshotResponse.java]: Please add "since" along with the description.

**Comment by koushik-das** [comment]
Please add/modify Marvin tests based on these scenarios.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by ProjectMoon** [comment]
Rebased to latest master. The comments will also be addressed soon, hopefully.

File: server/src/com/cloud/api/ApiResponseHelper.java (hunk 3 above)
Comment (ProjectMoon) [inline server/src/com/cloud/api/ApiResponseHelper.java]: I was going to do this, but then I wondered... what exactly would it return? A pair of `Long`s perhaps? For now I will leave it as-is, pending more discussion.

File: server/src/com/cloud/api/ApiResponseHelper.java (hunk 1 above)
Comment (ProjectMoon) [inline server/src/com/cloud/api/ApiResponseHelper.java]: It is used in ApiResponseHelper.

File: server/src/com/cloud/api/ApiResponseHelper.java (hunk 3 above)
Comment (ProjectMoon) [inline server/src/com/cloud/api/ApiResponseHelper.java]: My mistake. I guess it would only return the ID of the Guest OS... so from the interface level it'd be from volume ID -> Guest OS ID?

**Comment by ProjectMoon** [comment]
Code review issues have been addressed, with the exception of combining the Guest OS ID/VM Instance ID into a join method on VolumeDao. There was some compilation failure in the F5 plugin when I tried building locally. Will see if that happens here too.

**Comment by rhtyd** [comment]
@ProjectMoon thanks, can you add a JIRA id for this and also use that in your commit's PR.
@blueorangutan package

**Comment by ProjectMoon** [comment]
Updated to latest master, and now with a JIRA ticket.

https://issues.apache.org/jira/browse/CLOUDSTACK-9643

**Comment by rhtyd** [comment]
Thanks @ProjectMoon 
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
LGTM, the failure in Travis is caused by env issue and not the PR itself. Merging this based on review, and test results.
